### PR TITLE
Feat: Collateral Module

### PR DESCRIPTION
**Pull request by BLasan**
## Description
User Interfaces for Collateral Management

## Related issues and discussion
See [Issue-1357](https://issues.apache.org/jira/browse/FINERACT-1357)

## Screenshots, if any

- Collateral Product Page

Screenshot from 2021-09-15 23-12-51
Screenshot from 2021-09-15 23-13-41
Screenshot from 2021-09-15 23-13-45
Screenshot from 2021-09-15 23-14-55
Screenshot from 2021-09-15 23-15-27

- Client Collateral Page

Screenshot from 2021-09-15 23-16-29
Screenshot from 2021-09-15 23-16-39
Screenshot from 2021-09-15 23-17-56

- Loan Collateral Page

Screenshot from 2021-09-15 23-21-49
Screenshot from 2021-09-15 23-24-28

## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Validate the JS and HTML files with `grunt validate` to detect errors and potential problems in JavaScript code.

- [ ] Run the tests by opening `test/SpecRunner.html` in the browser to make sure you didn't break anything.

- [ ] If you have multiple commits please combine them into one commit by squashing them.

- [x] Read and understood the contribution guidelines at `community-app/Contributing.md`.

### CODE REVIEW

**Comment by BLasan** [comment]
@bharathcgowda Please review. Will attach the screen recording to the description

File: app/global-translations/locale-en.json [hunk 1]
```diff
@@ -213,6 +230,9 @@
   "label.anchor.editpermissions": "Edit Permissions",
   "#Inputs": "..",
   "label.input.name": "Name",
+  "label.input.unittype": "Unit Type",
+	"label.input.baseprice": "Base Price",
```
Comment (luckyman20) [inline app/global-translations/locale-en.json]: why extra space here?

File: app/global-translations/locale-en.json [hunk 2]
```diff
@@ -1298,13 +1322,14 @@
...
   "label.heading.addloancollateral": "Add Loan Collateral",
+  "label.heading.createcollateral": "Create Collateral",
   "label.heading.uploaddocument": "Upload Document",
   "label.heading.assignloanofficer": "Assign Loan Officer",
   "label.heading.guarantor": "Guarantor",
   "label.heading.createguarantor": "Create Guarantor",
   "label.heading.editloanaccount": "Edit Loan Account",
   "label.heading.editloanapplication": "Edit Loan Application",
-  "label.heading.collaterals": "Collateral",
+  "label.heading.collaterals1": "Collateral",
```
Comment (luckyman20) [inline app/global-translations/locale-en.json]: why did you rename this?

File: app/scripts/controllers/client/CreateClientCollateralController.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+                    scope.formData.unitType = scope.collateralData.unitType;
+                    scope.formData.collateralId = collateralId;
+                    scope.formData.quantity = 0.0;
+                    scope.formData.total = 0.0;
+                    scope.formData.totalCollateral = 0.0
+                    scope.disabled = false;
+                    console.log(scope.formData)
+                });
+
+                // resourceFactory.loanResource.get({resourceType: 'template', templateType: 'collateral', productId: loanProductId, fields: 'id,loanCollateralOptions'}, function (data) {
```
Comment (luckyman20) [inline app/scripts/controllers/client/CreateClientCollateralController.js]: Please check the commented code.

File: app/scripts/controllers/client/CreateClientCollateralController.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+                //     scope.collateralOptions = data.loanCollateralOptions || [];
+                // });
+
+            }
+
+            resourceFactory.collateralResource.getAllCollaterals(function (data) {
+                scope.collaterals = data;
+            });
+
+            // resourceFactory.loanCollateralTemplateResource.get({loanId: scope.loanId}, function (data) {
```
Comment (luckyman20) [inline app/scripts/controllers/client/CreateClientCollateralController.js]: Check this too.

File: app/scripts/controllers/client/EditClientCollateralController.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+                scope.collaterals = data;
+                scope.formData = {
+                    name: data.name,
+                    quantity: data.quantity,
+                    total: data.total,
+                    totalCollateral: data.totalCollateral
+                }
+            });
+
+            // resourceFactory.loanCollateralTemplateResource.get({loanId: scope.loanId}, function (data) {
```
Comment (luckyman20) [inline app/scripts/controllers/client/EditClientCollateralController.js]: This too.

File: app/scripts/controllers/client/ViewAllClientCollateralsController.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
+(function (module) {
+    mifosX.controllers = _.extend(module, {
+        ViewAllClientCollateralController: function (scope, resourceFactory, routeParams, location) {
+            scope.clientId = routeParams.id;
+            console.log(routeParams)
+            scope.routeTo = function (id) {
+                location.path('/clients/' + routeParams.id + '/viewclientcollateral/' + id);
+            };
+
+            // if (!scope.searchCriteria.collaterals) {
```
Comment (luckyman20) [inline app/scripts/controllers/client/ViewAllClientCollateralsController.js]: This too.

File: app/scripts/controllers/client/ViewAllClientCollateralsController.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+            //     scope.searchCriteria.collaterals = null;
+            //     scope.saveSC();
+            // }
+            //scope.filterText = scope.searchCriteria.charges || '';
+
+            // scope.onFilter = function () {
```
Comment (luckyman20) [inline app/scripts/controllers/client/ViewAllClientCollateralsController.js]: This too.

File: app/scripts/controllers/client/ViewClientCollateralController.js [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+            console.log(routeParams);
+
+            resourceFactory.clientcollateralResource.get({clientId: scope.clientId, collateralParamId: scope.collateralId}, function (data) {
+                scope.collateral = data;
+                console.log(data);
+                scope.loanTransactions = scope.collateral.loanTransactionData;
+                for (var i=0; i<scope.loanTransactions.length; i++){
+                	scope.loanTransactions[i].lastRepaymentDate.date = new Date(scope.loanTransactions[i].lastRepaymentDate.date);
+                }
+                console.log(scope.loanTransactions)
```
Comment (luckyman20) [inline app/scripts/controllers/client/ViewClientCollateralController.js]: I think this must be a temporary `console.log` statement. Please remove it.

File: app/scripts/controllers/client/ViewClientCollateralController.js [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+                        $uibModalInstance.close('delete');
+                        location.path('/viewclient/' + scope.clientId);
+                    });
+                };
+                $scope.cancel = function () {
+                    $uibModalInstance.dismiss('cancel');
+                };
+            };
+
+            // resourceFactory.loanCollateralTemplateResource.get({loanId: scope.loanId}, function (data) {
```
Comment (luckyman20) [inline app/scripts/controllers/client/ViewClientCollateralController.js]: This too.

File: app/scripts/controllers/client/ViewClientController.js [hunk 10]
```diff
@@ -95,6 +97,14 @@
 
             }
 
+            // collaterals
+
+            //resourceFactory.clientcollateralResource.getAllCollaterals({clientId: routeParams.id, currency: null}, function(data) {
```
Comment (luckyman20) [inline app/scripts/controllers/client/ViewClientController.js]: This too.

File: app/scripts/controllers/client/ViewClientController.js [hunk 11]
```diff
@@ -125,6 +132,14 @@
 
             }
 
+            scope.viewCollaterals=function()
+            {
+
```
Comment (luckyman20) [inline app/scripts/controllers/client/ViewClientController.js]: Please format this code block properly.

File: app/scripts/controllers/client/ViewClientController.js [hunk 12]
```diff
@@ -170,11 +185,18 @@
...
+            };
+
+            scope.routeToCollateral = function(id) {
+                location.path('/viewclient/' + routeParams.id + '/viewclientcollateral/' + id);
+            }
 
             scope.haveFile = [];
             resourceFactory.clientResource.get({clientId: routeParams.id}, function (data) {
                 scope.client = data;
+                console.log(scope.client);
```
Comment (luckyman20) [inline app/scripts/controllers/client/ViewClientController.js]: Another `console.log` statement.

File: app/scripts/controllers/loanAccount/EditLoanAccAppController.js [hunk 13]
```diff
@@ -79,6 +86,32 @@
...
+            }
+
+            scope.deleteCollateral = function (index) {
+                console.log(index);
+                console.log(scope.collaterals[index]);
+                scope.collateralId = scope.collaterals[index].collateralId;
+                scope.collateralObject = scope.collateralAddedDataArray.filter((collateral) => collateral.collateralId == scope.collateralId)[0];
+                scope.collateralsData.push(scope.collateralObject);
+                scope.collaterals.splice(index, 1);
+
```
Comment (luckyman20) [inline app/scripts/controllers/loanAccount/EditLoanAccAppController.js]: Blank line

File: app/scripts/controllers/loanAccount/EditLoanAccAppController.js [hunk 14]
```diff
@@ -262,18 +295,18 @@
...
 
-            scope.addCollateral = function () {
-                if (scope.collateralFormData.collateralIdTemplate && scope.collateralFormData.collateralValueTemplate) {
-                    scope.collaterals.push({type: scope.collateralFormData.collateralIdTemplate.id, name: scope.collateralFormData.collateralIdTemplate.name, value: scope.collateralFormData.collateralValueTemplate, description: scope.collateralFormData.collateralDescriptionTemplate});
-                    scope.collateralFormData.collateralIdTemplate = undefined;
-                    scope.collateralFormData.collateralValueTemplate = undefined;
-                    scope.collateralFormData.collateralDescriptionTemplate = undefined;
-                }
-            };
+            // scope.addCollateral = function () {
```
Comment (luckyman20) [inline app/scripts/controllers/loanAccount/EditLoanAccAppController.js]: Commented code.

File: app/scripts/controllers/loanAccount/EditLoanAccAppController.js [hunk 15]
```diff
@@ -262,18 +295,18 @@
...
+            //         scope.collateralFormData.collateralIdTemplate = undefined;
+            //         scope.collateralFormData.collateralValueTemplate = undefined;
+            //         scope.collateralFormData.collateralDescriptionTemplate = undefined;
+            //     }
+            // };
 
-            scope.deleteCollateral = function (index) {
-                scope.collaterals.splice(index, 1);
-            };
+            // scope.deleteCollateral = function (index) {
```
Comment (luckyman20) [inline app/scripts/controllers/loanAccount/EditLoanAccAppController.js]: Commented code.

File: app/scripts/controllers/loanAccount/EditLoanAccAppController.js [hunk 16]
```diff
@@ -349,12 +382,25 @@
                     }
                 }
 
-                scope.formData.collateral = [];
+                // scope.formData.collateral = [];
```
Comment (luckyman20) [inline app/scripts/controllers/loanAccount/EditLoanAccAppController.js]: Too much commented code in this file.

File: app/scripts/controllers/loanAccount/NewLoanAccAppController.js [hunk 17]
```diff
@@ -283,17 +294,37 @@
                 }
             };
 
+            // scope.addCollateral = function () {
```
Comment (luckyman20) [inline app/scripts/controllers/loanAccount/NewLoanAccAppController.js]: Commented code.

File: app/scripts/controllers/loanAccount/ViewLoanCollateralDataController.js [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+    mifosX.controllers = _.extend(module, {
+        ViewLoanCollateralDataController: function (scope, resourceFactory, routeParams, location, $uibModal) {
+
+            scope.loanId = routeParams.id;
+            scope.collateralId = routeParams.collateralId;
+            //scope.showEditButtons = routeParams.status == 'Submitted and pending approval' ? true : false;
+            resourceFactory.loancollateralResource.get({collateralId: scope.collateralId}, function (data) {
+                scope.collateral = data;
+            });
+            // scope.deleteCollateral = function () {
```
Comment (luckyman20) [inline app/scripts/controllers/loanAccount/ViewLoanCollateralDataController.js]: Commented code.

File: app/scripts/controllers/loanAccount/ViewLoanDetailsController.js [hunk 19]
```diff
@@ -152,6 +152,7 @@
 
             resourceFactory.LoanAccountResource.getLoanAccountDetails({loanId: routeParams.id, associations: 'all',exclude: 'guarantors,futureSchedule'}, function (data) {
                 scope.loandetails = data;
+                console.log(scope.loandetails);
```
Comment (luckyman20) [inline app/scripts/controllers/loanAccount/ViewLoanDetailsController.js]: Another `console.log` statement.

**Review by luckyman20 — COMMENTED** [review]
Please check all the commented out code (I pointed out some but there is more) and remove console.log statement too.

**Comment by BLasan** [comment]
@luckyman20 Thanks for the review. Will refine as suggested

File: app/global-translations/locale-en.json (hunk 1 above)
Comment (BLasan) [reply app/global-translations/locale-en.json]: Pushed. But this is not getting updated
Screenshot from 2021-09-10 21-11-04

Comment (BLasan) [reply app/global-translations/locale-en.json]: Done

File: app/global-translations/locale-en.json (hunk 2 above)
Comment (BLasan) [reply app/global-translations/locale-en.json]: Fixed

File: app/scripts/controllers/client/CreateClientCollateralController.js (hunk 3 above)
Comment (BLasan) [reply app/scripts/controllers/client/CreateClientCollateralController.js]: Removed

File: app/scripts/controllers/client/ViewClientCollateralController.js (hunk 8 above)
Comment (BLasan) [reply app/scripts/controllers/client/ViewClientCollateralController.js]: Removed unwanted console logs

File: app/scripts/controllers/client/ViewClientController.js (hunk 11 above)
Comment (BLasan) [reply app/scripts/controllers/client/ViewClientController.js]: Done

File: app/scripts/controllers/loanAccount/EditLoanAccAppController.js (hunk 13 above)
Comment (BLasan) [reply app/scripts/controllers/loanAccount/EditLoanAccAppController.js]: Done

**Comment by BLasan** [comment]
@luckyman20 Please review

File: app/global-translations/locale-en.json [hunk 20]
```diff
@@ -213,6 +230,9 @@
   "label.anchor.editpermissions": "Edit Permissions",
   "#Inputs": "..",
   "label.input.name": "Name",
+  "label.input.unittype": "Unit Type",
+	"label.input.pcttobase": "Percentage To Base",
+	"label.input.baseprice": "Base Price",
```
Comment (luckyman20) [inline app/global-translations/locale-en.json:235]:
```suggestion
"label.input.baseprice": "Base Price",
```

File: app/global-translations/locale-en.json [hunk 21]
```diff
@@ -213,6 +230,9 @@
   "label.anchor.editpermissions": "Edit Permissions",
   "#Inputs": "..",
   "label.input.name": "Name",
+  "label.input.unittype": "Unit Type",
+	"label.input.pcttobase": "Percentage To Base",
```
Comment (luckyman20) [inline app/global-translations/locale-en.json:234]:
```suggestion
"label.input.pcttobase": "Percentage To Base",
```

You can directly apply these suggestions.

File: app/global-translations/locale-en.json [hunk 22]
```diff
@@ -1298,13 +1322,13 @@
...
   "label.heading.addloancharge": "Add Loan Charge",
   "label.heading.addloancollateral": "Add Loan Collateral",
+  "label.heading.createcollateral": "Create Collateral",
   "label.heading.uploaddocument": "Upload Document",
   "label.heading.assignloanofficer": "Assign Loan Officer",
   "label.heading.guarantor": "Guarantor",
   "label.heading.createguarantor": "Create Guarantor",
   "label.heading.editloanaccount": "Edit Loan Account",
   "label.heading.editloanapplication": "Edit Loan Application",
-  "label.heading.collaterals": "Collateral",
```
Comment (luckyman20) [inline app/global-translations/locale-en.json:1307]: Is this string not used anywhere?

File: app/scripts/controllers/client/ViewClientController.js [hunk 23]
```diff
@@ -17,6 +17,8 @@
             scope.updateDefaultSavings = false;
             scope.charges = [];
 
+            scope.collaterals = [];
```
Comment (luckyman20) [inline app/scripts/controllers/client/ViewClientController.js:20]: Is this array required later in the file?

**Review by luckyman20 — COMMENTED** [review]
Code quality otherwise seems good. I will need to run a few edge cases and then it will be good to merge!

File: app/global-translations/locale-en.json (hunk 21 above)
Comment (BLasan) [reply app/global-translations/locale-en.json:234]: When I squash and re-push these changes won't get applied properly.  In my local branch these things are fixed

File: app/global-translations/locale-en.json (hunk 22 above)
Comment (BLasan) [reply app/global-translations/locale-en.json:1307]: Moved to line 5

File: app/scripts/controllers/client/ViewClientController.js (hunk 23 above)
Comment (BLasan) [reply app/scripts/controllers/client/ViewClientController.js:20]: Yes. In lines 179,180,

File: app/global-translations/locale-en.json [hunk 24]
```diff
@@ -213,6 +230,9 @@
   "label.anchor.editpermissions": "Edit Permissions",
   "#Inputs": "..",
   "label.input.name": "Name",
+  "label.input.unittype": "Unit Type",
+"label.input.pcttobase": "Percentage To Base",
```
Comment (luckyman20) [inline app/global-translations/locale-en.json]: This isn't aligned.

File: app/scripts/controllers/client/CreateClientCollateralController.js [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+    mifosX.controllers = _.extend(module, {
+        CreateClientCollateralController: function (scope, resourceFactory, routeParams, location) {
+
+            scope.formData = {};
+            scope.clientId = routeParams.id;
+            scope.collateralData = {};
+            scope.disable = true;
+            scope.collateralDataRequestBody = {};
+            scope.collateralId;
+            console.log(routeParams);
```
Comment (luckyman20) [inline app/scripts/controllers/client/CreateClientCollateralController.js]: `console.log` statement to be removed

File: app/scripts/controllers/client/CreateClientCollateralController.js [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+                    scope.formData.type = scope.collateralData.quality;
+                    scope.formData.basePrice = scope.collateralData.basePrice;
+                    scope.formData.pctToBase = scope.collateralData.pctToBase;
+                    scope.formData.unitType = scope.collateralData.unitType;
+                    scope.formData.collateralId = collateralId;
+                    scope.formData.quantity = 0.0;
+                    scope.formData.total = 0.0;
+                    scope.formData.totalCollateral = 0.0
+                    scope.disabled = false;
+                    console.log(scope.formData)
```
Comment (luckyman20) [inline app/scripts/controllers/client/CreateClientCollateralController.js]: `console.log` statement to be removed

File: app/scripts/controllers/client/CreateClientCollateralController.js [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+                scope.collaterals = data;
+            });
+
+            scope.cancel = function () {
+                location.path('/viewclient/' + scope.clientId);
+            };
+
+            scope.submit = function () {
+                this.formData.locale = scope.optlang.code;
+                console.log(this.formData);
```
Comment (luckyman20) [inline app/scripts/controllers/client/CreateClientCollateralController.js]: `console.log` statement to be removed

File: app/scripts/controllers/client/EditClientCollateralController.js [hunk 28]
```diff
@@ -0,0 +1,40 @@
+(function (module) {
+    mifosX.controllers = _.extend(module, {
+        EditClientCollateralController: function (scope, resourceFactory, routeParams, location) {
+
+            scope.formData = {};
+            scope.clientId = routeParams.id;
+            scope.collateralId = routeParams.collateralId;
+            scope.collateralDataRequestBody = {};
+            console.log(routeParams);
```
Comment (luckyman20) [inline app/scripts/controllers/client/EditClientCollateralController.js]: `console.log` statement to be removed

File: app/scripts/controllers/client/ViewAllClientCollateralsController.js [hunk 29]
```diff
@@ -0,0 +1,18 @@
+(function (module) {
+    mifosX.controllers = _.extend(module, {
+        ViewAllClientCollateralController: function (scope, resourceFactory, routeParams, location) {
+            scope.clientId = routeParams.id;
+            scope.routeTo = function (id) {
+                location.path('/clients/' + routeParams.id + '/viewclientcollateral/' + id);
+            };
+
+            scope.CollateralPerPage =15;
```
Comment (luckyman20) [inline app/scripts/controllers/client/ViewAllClientCollateralsController.js]:
```suggestion
            scope.CollateralPerPage = 15;
```

File: app/scripts/controllers/loanAccount/NewLoanAccAppController.js [hunk 30]
```diff
@@ -284,16 +293,23 @@
...
+                scope.collateralFormData.total = scope.collateralFormData.quantity * scope.collateralObject.basePrice;
+                scope.collateralFormData.totalCollateral = scope.collateralFormData.total * scope.collateralObject.pctToBase / 100.0;
+            }
+
             scope.deleteCollateral = function (index) {
+                scope.collateralId = scope.collaterals[index].collateralId;
+                scope.collateralObject = scope.collateralAddedDataArray.filter((collateral) => collateral.collateralId == scope.collateralId)[0];
+                scope.collateralsData.push(scope.collateralObject);
                 scope.collaterals.splice(index, 1);
+
```
Comment (luckyman20) [inline app/scripts/controllers/loanAccount/NewLoanAccAppController.js]: Blank line added.

File: app/scripts/controllers/loanAccount/ViewLoanCollateralDataController.js [hunk 31]
```diff
@@ -0,0 +1,16 @@
+(function (module) {
+    mifosX.controllers = _.extend(module, {
+        ViewLoanCollateralDataController: function (scope, resourceFactory, routeParams, location, $uibModal) {
+
+            scope.loanId = routeParams.id;
+            scope.collateralId = routeParams.collateralId;
+            //scope.showEditButtons = routeParams.status == 'Submitted and pending approval' ? true : false;
```
Comment (luckyman20) [inline app/scripts/controllers/loanAccount/ViewLoanCollateralDataController.js]: Commented out line.

**Comment by BLasan** [comment]
@luckyman20 Added all the requested changes

**Comment by ikimbra** [comment]
@luckyman20 @bharathcgowda @BLasan @francisguchie @rrpawar96  we are no add able to add simple loan collateral because of the absence of this module. I created the issue on Jira [here](https://issues.apache.org/jira/browse/FINERACT-1749) 

Please merge this as soon as possible since the collateral functionality is completely broken.

**Comment by eroemma** [comment]
This issue is critical since collateral is broken.

**Comment by francisguchie** [comment]
@eroemma @ikimbra 
Let us thoroughly do a User / functionality test and provide feed back asap to @BLasan @luckyman20

**Comment by francisguchie** [comment]
@luckyman20  @BLasan  
The tests i have done provide good results. this is a good enhancement. Kindly merge
